File: abs-divide/src/main/java/com/problem/absdiv/AbsDiv.java
package com.problem.absdiv;

import static java.lang.Math.max;
import static java.lang.Math.abs;

public class AbsDiv {
	public static int solution(int... A) {
		int[] B = new int[A.length];
		
		int max = Integer.MIN_VALUE;
		for (int i = A.length - 1; i >= 0; --i)
			B[i] = max = max(max, A[i]);
		
		max = Integer.MIN_VALUE;
		for (int i = 0; i < A.length; ++i)
			A[i] = max = max(max, A[i]);
		
		max = Integer.MIN_VALUE;
		for (int i = 0; i < A.length - 1; ++i)
			max = max(max, abs(A[i] - B[i + 1]));
		
		return max;
	}
}

File: abs-divide-scala/src/main/scala/com/problem/absdiv/AbsDiv.scala
package com.problem.absdiv

object AbsDiv {

  def solve(a: Array[Int]): Int = {
    val L = a.size
    val b = new Array[Int](L)

    b(L - 1) = a(L - 1) // only base case, to avoid copying all array

    for (i <- (L - 2) to 0 by -1)
      b(i) = b(i + 1).max(a(i)) // purposedly reading current value from 'a'

    for (i <- 1 until L)
      a(i) = a(i - 1).max(a(i))

    (0 until (L - 1))
      .map { i => (a(i) - b(i + 1)).abs }
      .max
  }
}
